example/contrib/socks-client/socks/impl/handshake.hpp

```diff
@@ -0,0 +1,598 @@
+            auto status = read<uint8_t>(resp);
+
+            if (version != 0x01) // auth version.
+            {
+                ec = error::socks_unsupported_authentication_version;
+                break;
+            }
+
+            if (status != 0x00)
+            {
+                ec = error::socks_authentication_error;
+                break;
+            }
+            BOOST_FALLTHROUGH;
+        }
+        case 5:
+        {
+            step_ = 6;
+
+            request.consume(request.size());
+            std::size_t bytes_to_write = 7 + hostname_.size();
+            auto req = static_cast<char*>(
+                request.prepare(std::max<std::size_t>(bytes_to_write, 22)).data());
+
+            write<uint8_t>(detail::SOCKS_VERSION_5, req); // SOCKS VERSION 5.
+            write<uint8_t>(detail::SOCKS_CMD_CONNECT, req); // CONNECT command.
+            write<uint8_t>(0, req); // reserved.
+
+            if (use_hostname_)
+            {
+                write<uint8_t>(detail::SOCKS5_ATYP_DOMAINNAME, req); // atyp, domain name.
+                BOOST_ASSERT(hostname_.size() <= 255);
+                write<uint8_t>(static_cast<int8_t>(hostname_.size()), req);    // domainname size.
+                std::copy(hostname_.begin(), hostname_.end(), req);    // domainname.
+                req += hostname_.size();
+                write<uint16_t>(port_, req);    // port.
+            }
+            else
+            {
+                auto endp = net::ip::make_address(hostname_, ec);
+                if (ec)
+                    break;
+
+                if (endp.is_v4())
+                {
+                    write<uint8_t>(detail::SOCKS5_ATYP_IPV4, req); // ipv4.
+                    write<uint32_t>(endp.to_v4().to_uint(), req);
+                    write<uint16_t>(port_, req);
+                    bytes_to_write = 10;
+                }
+                else
+                {
+                    write<uint8_t>(detail::SOCKS5_ATYP_IPV6, req); // ipv6.
+                    auto bytes = endp.to_v6().to_bytes();
+                    std::copy(bytes.begin(), bytes.end(), req);
+                    req += 16;
+                    write<uint16_t>(port_, req);
+                    bytes_to_write = 22;
+                }
+            }
+
+            request.commit(bytes_to_write);
+            return net::async_write(
+                stream_,
+                request,
+                std::move(*this));
+        }
+        case 6:
+        {
+            step_ = 7;
+
+            std::size_t bytes_to_read = 10;
+            response.consume(response.size());
+            return net::async_read(
+                stream_,
+                response,
+                net::transfer_exactly(bytes_to_read),
+                std::move(*this));
+        }
+        case 7:
+        {
+            step_ = 8;
+
+            auto resp = static_cast<const char*>(response.data().data());
+            auto ver = read<uint8_t>(resp);
```

Comment:
All this code and other business logic needs to be factored out into individual implementation-specific functions.

Reply:
done in socks 4. pending for socks 5